_blogposts/2025-11-11-introducing-rewatch.mdx

```diff
@@ -0,0 +1,266 @@
+```
+
+The internal implementation changed, but the public API (the `make` function signature) stayed the same. The `.cmi` file is identical before and after.
+
+**With the old build system:**
+
+- Button.res changes → recompile Button
+- Check all dependents of Button → recompile them too
+- Check all their dependents → recompile those as well
+- Result: potentially dozens of modules recompiled
+
+**With ReWatch:**
+
+- Button.res changes → recompile Button
+- Check Button.cmi hash → unchanged
+- Skip recompiling dependents
+- Result: one module recompiled
+
+This is the key innovation. ReWatch doesn't just track which modules depend on each other. It understands when those dependencies actually matter.
+
+### When Recompilation is Necessary
+
+Of course, when you do change a public interface, ReWatch correctly identifies all affected modules:
+
+```rescript
+// Button.res - Adding a new parameter
+let make = (~size, ~onClick, ~disabled=false) => {
+  // ...
+}
+```
+
+Now the `.cmi` file changes because the function signature changed. ReWatch detects this and recompiles all modules that use `Button.make`. This is the correct behaviour, but it only happens when truly necessary.
+
+### Explicit Interface Files for Maximum Control
+
+If you want even more control over your module's public interface, you can use explicit `.resi` files:
+
+```rescript
+// Button.resi
+type size = Small | Medium | Large
+let make: (~size: size, ~onClick: unit => unit) => Jsx.element
+let defaultSize: size
+```
+
+With an explicit interface file, the `.cmi` is generated from the `.resi` file. Any changes to `Button.res` that don't affect the `.resi` contract will never trigger dependent recompilations. This gives you maximum build performance for frequently-changed modules.
+
+**Bonus for React developers:** Using `.resi` files for your React components has another benefit. During development, when you modify the component's internal implementation without changing the interface, React's [Fast Refresh](https://www.gatsbyjs.com/docs/reference/local-development/fast-refresh/) can preserve component state more reliably. Combined with ReWatch's intelligent rebuilding, this makes for an exceptionally smooth development experience.
+
+### Faster Module Resolution with Flat Directory Layout
+
+ReWatch employs another clever optimization for module resolution. When building your project, it copies all source files to a flat directory structure at `lib/ocaml/`. Instead of maintaining the original nested directory structure, every module ends up in one place.
+
+Think of it like organizing a library. The old approach was like having books scattered across multiple rooms and floors. To find a specific book, you'd need to check each room. ReWatch's approach is like putting all the books in one room. Finding what you need is instant.
+
+**Why this matters:**
+
+- Module lookup becomes a single directory operation
+- The filesystem cache is more effective when files are adjacent
```

Comment:
Just because _someone_ might comment on it - file systems struggle when there are _too many_ files in a folder as well. But, since Rewatch takes ownership over where the files end up, sharding the layout later on as a perf optimization can be done entirely by Rewatch, and transparently to the user.